app-backend/backsplash-core/src/main/scala/com/rasterfoundry/backsplash/BacksplashImage.scala

```diff
@@ -166,6 +170,195 @@ final case class BacksplashGeotiff(
+  val tags = Map(
+    "imageName" -> (imageName getOrElse ""),
+    "imageId" -> s"$imageId",
+    "subsetBands" -> subsetBands.mkString(","),
+    "prefix" -> prefix,
+    "readType" -> "LandsatMultitiff"
+  )
+
+  def getUri(band: Int): Option[String] = imageName map { name =>
+    s"$prefix/${name}_B${band + 1}.TIF"
+  }
+
+  def getBandRasterSource(i: Int,
+                          context: TracingContext[IO]): IO[RasterSource] = {
+    val uri = getUri(i) getOrElse { "" }
+    val rsTags = tags.combine(Map("uri" -> uri))
+    context.childSpan("getBandRasterSource", rsTags) use { _ =>
+      logger.debug(s"Using GDAL Raster Source: ${uri}")
+      // Do not bother caching - let GDAL internals worry about that
+      val rasterSource = GDALRasterSource(URLDecoder.decode(uri, "UTF-8"))
+      IO {
+        metadata.noDataValue match {
+          case Some(nd) =>
+            rasterSource.interpretAs(DoubleUserDefinedNoDataCellType(nd))
+          case _ =>
+            rasterSource
+        }
+
+      }
+    }
+  }
+
+  def selectBands(bands: List[Int]) = this.copy(subsetBands = bands)
+}
+
+sealed abstract class MultiTiffImage[F[_]: Monad, G[_]](
+    implicit F: Parallel[F, G],
+    sync: Sync[F],
+    mode: Mode[F],
+    Err: MonadError[F, Throwable]
+) extends BacksplashImage[F]
+    with LazyLogging {
+
+  implicit val tileCache = Cache.tileCache
+  implicit val rasterSourceCache = Cache.rasterSourceCache
+
+  val tags: Map[String, String]
+
+  def getRasterSource(context: TracingContext[F]): F[RasterSource] =
+    context.childSpan("getRasterSource", tags) use { child =>
+      getBandRasterSource(subsetBands.headOption getOrElse 0, child)
+    }
+
+  /** Get a single band raster source for one band of this image
+    *
+    * MultiTiff image assumes that you have a single scene split up over several
+    * single band tiffs, e.g., how Landast 8 and Sentinel-2 are stored on AWS.
+    * This assumption will cause some things to break if you end up trying to make
+    * a MultiTiffImage out of a multi-band tiff, e.g., you might think you're going
+    * to color correct one way and have something else happen entirely.
+    */
+  def getBandRasterSource(i: Int, context: TracingContext[F]): F[RasterSource]
+
+  def getBandRasterSources(
+      bs: NonEmptyList[Int],
+      context: TracingContext[F]
+  ): F[NonEmptyList[RasterSource]] =
+    bs parTraverse { band =>
+      getBandRasterSource(band, context)
+    }
+
+  def getHistogram(context: TracingContext[F]): F[Array[Histogram[Double]]] =
+    context.childSpan("readHistogramFromSource", Map.empty) use { child =>
+      for {
+        sources <- subsetBands.toNel match {
+          case Some(bs) => getBandRasterSources(bs, child)
+          case None =>
+            Err.raiseError(
+              RequirementFailedException("Must request at least one band"))
+        }
+        zoomedOutExtents <- sources parTraverse { rs =>
```

Comment:
I was looking at traces - and saw that there was a span here that didn't have a lot of information about what was taking so long for it.

![image](https://user-images.githubusercontent.com/898060/65044238-87b2b300-d92a-11e9-975f-678389d39b06.png)

I added a few more children spans - specifically inside the `parTraverse` - but nothing really accounted for it:

![image](https://user-images.githubusercontent.com/898060/65044400-d2342f80-d92a-11e9-91b1-dae84fe4e881.png)

Not sure we're going to get to the bottom of it here, but might be a good idea to instrument those reads individually.

Reply:
I don't think we're going to get to the bottom of it here either. I added another trace for constructing the multibandtile in f136686, so that we can see where the end of image reading is. I don't know what my guess is. I'd like to refactor to use fibers and see what changes but I'd like to hold off on that in this PR. I think it might be worthwhile to write up our tracing adventures (once x-ray is in) in a blog post since we're now actively using it for debugging, then drop it in cats-effect gitter and see who bites.

Reply:
Also, this is what it looks like with the mbt construction trace:

![image](https://user-images.githubusercontent.com/5702984/65050774-9189e580-d92d-11e9-90de-5dd7a0762e03.png)

At that point in `MosaicImplicits` it's ready to join the hists fiber, and yet... we still have a huge wait before that fiber join is complete despite all the work being done... maybe. It just occurred to me that maybe the histogram calculation itself is waiting on the join call? I'm gonna trace that as well.